Open settings submenu links in new tabs

vidushigupta0607 opened the pull request:
**Problem:**
Several external links on WordPress admin Settings pages—specifically General, Discussion, and Permalink—currently open in the same window. This creates multiple user experience issues:

Users may lose unsaved changes if they click an external link without saving.
Users may expect links to open in a new tab (like Edit Profile or Help links) and unintentionally navigate away.
Inconsistent link behavior across admin pages causes confusion and a fragmented interface.

**Root Cause:**
Some documentation/help links in the Settings pages lack target="_blank", making the UI inconsistent and unpredictable.

**Solution:**
Update affected external documentation/help links to include target="_blank", so they open in a new tab.
Add visual and screen-reader indicators for new-tab links to improve accessibility:
- External link icon (Dashicons)
- aria-label text for screen readers
- Keep internal admin links navigating between Settings pages in the same tab, and add a JS warning for unsaved changes to prevent data loss.

**Benefits:**
- Prevents accidental data loss on Settings pages.
- Ensures consistent and predictable link behavior.
- Improves accessibility and UX for all users.

Trac Ticket (New/Main): [#64623](https://core.trac.wordpress.org/ticket/64623)

Related Ticket (Old): [#23432](https://core.trac.wordpress.org/ticket/23432)

vidushigupta0607 committed:
Open settings submenu links in new tabs

vidushigupta0607 commented:
The account has been linked now!

shail-mehta commented on src/wp-admin/options-general.php:
```diff
@@ -250,7 +250,7 @@ class="<?php echo esc_attr( $classes_for_button ); ?>"
 		printf(
 			/* translators: %s: Documentation URL. */
-			__( 'Enter the same address here unless you <a href="%s">want your site home page to be different from your WordPress installation directory</a>.' ),
+			__( 'Enter the same address here unless you <a href="%s" target=_blank>want your site home page to be different from your WordPress installation directory</a>.' ),
```
```suggestion
			__( 'Enter the same address here unless you <a href="%s" target="_blank">want your site home page to be different from your WordPress installation directory</a>.' ),
```

vidushigupta0607 replied:
The PR has been updated as per the new ticket for the issue

vidushigupta0607 committed:
Settings: Add unsaved changes warning and fix link accessibility.

Implement a JS-based beforeunload warning that alerts users when they attempt to navigate away from settings pages with unsaved changes. The warning:
- Only triggers when actual form changes exist (via serialize comparison)
- Handles all navigation scenarios (internal links, back/forward, tab close)
- Does NOT fire for new-tab links (which don't unload the current page)
- Suppresses warning on intentional form submission (Save Changes)
- Preserves browser bfcache by lazy-attaching beforeunload only after first user change

Additionally, fix inconsistent link accessibility across settings pages:
- Remove target="_blank" from internal admin links (e.g., moderation queue in Discussion Settings)
  to restore user control and rely on the new unsaved-changes warning for protection
- Add target="_blank" + accessibility indicators to external documentation/preview links
  following WordPress canonical pattern (visual icon + screen-reader text)
- Add rel="noopener noreferrer" for security on all new-tab links
- Ensure consistent behavior across General, Discussion, Reading, Writing, Permalink, and Privacy pages

Files modified:
- src/js/_enqueues/admin/settings.js: New module with beforeunload handler and lazy attachment
- src/wp-admin/options-head.php: Enqueue settings.js on all settings pages
- src/wp-includes/script-loader.php: Register settings script handle
- src/wp-admin/options-*.php: Update 9 link instances across 6 settings pages
- Gruntfile.js: Add build entries for new settings.js module
- tests/qunit/: Add QUnit tests for beforeunload behavior

Props: Accessibility review team, WordPress core team
Fixes: #64623 (Prevent losing data when clicking links on Settings pages)

vidushigupta0607 committed:
Tests: Fix settings.js QUnit tests to properly initialize module behavior.

The settings.js module initializes at jQuery DOM ready before test forms
are created in QUnit beforeEach hooks. This caused tests to fail because
the module exited early when it couldn't find the settings form.

Fixed by replicating the settings module's initialization logic in the
test's beforeEach hook after creating the test form. This ensures the
lazy attachment handlers are properly attached to test forms.

Also simplified the lazy attachment test to directly check that
beforeunload returns undefined before changes and a message after,
rather than attempting to count handler invocations.

All 466 QUnit tests now pass.

vidushigupta0607 committed:
Fix: Correct indentation in options-general.php line 579.

The closing PHP tag was missing a tab, causing a Generic.WhiteSpace.ScopeIndent
PHPCS error. Added proper tab indentation to match surrounding code structure.
